src/aspire/source/__init__.py

```diff
@@ -114,51 +116,49 @@ def states(self):
 
     @states.setter
     def states(self, values):
-        return self.set_metadata('_rlnClassNumber', values)
+        self.set_metadata('_rlnClassNumber', values)
 
     @property
-    def filters(self):
-        return self.get_metadata('__filter')
-
-    @filters.setter
-    def filters(self, values):
-        self.set_metadata('__filter', values)
-        if values is None:
-            new_values = np.nan
+    def filter_indices(self):
+        return self.get_metadata('__filter_indices')
+
+    @filter_indices.setter
+    def filter_indices(self, indices):
+        # create metadata of filters for all images
+        if indices is None:
+            filter_values = np.nan
         else:
-            new_values = np.array([(
-                getattr(f, 'voltage', np.nan),
-                getattr(f, 'defocus_u', np.nan),
-                getattr(f, 'defocus_v', np.nan),
-                getattr(f, 'defocus_ang', np.nan),
-                getattr(f, 'Cs', np.nan),
-                getattr(f, 'alpha', np.nan)
-            ) for f in values])
+            filter_values = np.array([(
+                getattr(self.unique_filters[i], 'voltage', np.nan),
+                getattr(self.unique_filters[i], 'defocus_u', np.nan),
+                getattr(self.unique_filters[i], 'defocus_v', np.nan),
+                getattr(self.unique_filters[i], 'defocus_ang', np.nan),
+                getattr(self.unique_filters[i], 'Cs', np.nan),
+                getattr(self.unique_filters[i], 'alpha', np.nan)
+            ) for i in indices])
 
         self.set_metadata(
-            ['_rlnVoltage', '_rlnDefocusU', '_rlnDefocusV', '_rlnDefocusAngle', '_rlnSphericalAberration', '_rlnAmplitudeContrast'],
-            new_values
+            ['_rlnVoltage', '_rlnDefocusU', '_rlnDefocusV', '_rlnDefocusAngle',
+             '_rlnSphericalAberration', '_rlnAmplitudeContrast'],
+            filter_values
         )
-
-    @property
-    def filter_indices(self):
-        return self.get_metadata('__filter_indices')
+        self.set_metadata(['__filter_indices'], indices)
 
     @property
     def offsets(self):
         return self.get_metadata(['_rlnOriginX', '_rlnOriginY'], default_value=0.)
 
     @offsets.setter
     def offsets(self, values):
-        return self.set_metadata(['_rlnOriginX', '_rlnOriginY'], values)
+        self.set_metadata(['_rlnOriginX', '_rlnOriginY'], values)
```

Comment:
why remove the `return`? If `self.set_metadata` chooses to return `None` (as it does) that's it's responsibility. This setter is simply relaying the call, not intercepting the return value.

Reply:
Pycharm complained no need to return for setter. Put it back in new version since it is preferred.